File: scala/src/Rational.scala
import scala.language.implicitConversions
import scala.math.BigInt

class Rational(num: BigInt, den: BigInt = 1) extends Ordered[Rational] {
  private def normalize(num: BigInt, den: BigInt) = {
    val g = if (den.signum > 0) num.gcd(den) else if (den.signum < 0) -num.gcd(den) else BigInt(1)
    (num / g, den / g)
  }

  val (n, d) = normalize(num, den)

  def +(that: Rational) = new Rational(n * that.d + that.n * d, d * that.d)

  def -(that: Rational) = new Rational(n * that.d - that.n * d, d * that.d)

  def *(that: Rational) = new Rational(n * that.n, d * that.d)

  def /(that: Rational) = new Rational(n * that.d, d * that.n)

  def unary_- = new Rational(-n, d)

  def abs = new Rational(n.abs, d)

  def signum = n.signum

  def doubleValue = n.doubleValue / d.doubleValue

  def longValue = n.longValue / d.longValue

  override def compare(that: Rational) = (n * that.d).compare(that.n * d)

  def max(that: Rational) = if (this > that) this else that

  def min(that: Rational) = if (this < that) this else that

  override def equals(that: Any) = that match {
    case that: Rational => n == that.n && d == that.d
    case _ => false
  }

  override def hashCode = n.hashCode() * 31 + d.hashCode()

  override def toString = n + "/" + d
}

object Rational {
  final val Zero = new Rational(0)
  final val One = new Rational(1)
  final val PositiveInfinity = new Rational(1, 0)
  final val NegativeInfinity = new Rational(-1, 0)

  implicit def int2Rational(i: Int): Rational = new Rational(i)

  implicit def long2Rational(l: Long): Rational = new Rational(l)

  // Usage example
  def main(args: Array[String]) {
    val a = 1 / new Rational(3)
    val b = new Rational(1, 6)
    val c = new Rational(1, 2)
    assert(c == a + b)
  }
}

File: kotlin/Rational.kt
import java.math.BigInteger

class Rational(n: BigInteger, d: BigInteger = BigInteger.ONE) : Comparable<Rational> {
    val num: BigInteger
    val den: BigInteger

    init {
        val gcd = n.gcd(d)
        val g = if (d.signum() > 0) gcd else if (d.signum() < 0) gcd.negate() else BigInteger.ONE
        num = n.divide(g)
        den = d.divide(g)
    }

    constructor(num: Long, den: Long = 1) : this(BigInteger.valueOf(num), BigInteger.valueOf(den))

    operator fun plus(r: Rational): Rational = Rational(num.multiply(r.den).add(r.num.multiply(den)), den.multiply(r.den))

    operator fun minus(r: Rational): Rational = Rational(num.multiply(r.den).subtract(r.num.multiply(den)), den.multiply(r.den))

    operator fun times(r: Rational): Rational = Rational(num.multiply(r.num), den.multiply(r.den))

    operator fun div(r: Rational): Rational = Rational(num.multiply(r.den), den.multiply(r.num))

    operator fun unaryMinus(): Rational = Rational(num.negate(), den)

    fun inverse(): Rational = Rational(den, num)

    fun abs(): Rational = Rational(num.abs(), den)

    fun signum(): Int = num.signum()

    fun doubleValue(): Double = num.toDouble() / den.toDouble()

    fun longValue(): Long = num.toLong() / den.toLong()

    override fun compareTo(other: Rational): Int = num.multiply(other.den).compareTo(other.num.multiply(den))

    override fun equals(other: Any?): Boolean = num == (other as Rational).num && den == other.den

    override fun hashCode(): Int = num.hashCode() * 31 + den.hashCode()

    override fun toString(): String = "$num/$den"

    companion object {
        val ZERO = Rational(0)
        val ONE = Rational(1)
        val POSITIVE_INFINITY = Rational(1, 0)
        val NEGATIVE_INFINITY = Rational(-1, 0)
    }
}

// Usage example
fun main() {
    val a = Rational(1, 3)
    val b = Rational(1, 6)
    val s1 = Rational(1, 2)
    val s2 = a + b
    println(true == (s1 == s2))
    println(a * b / b - a)
    println(a * Rational.ZERO)
}
